### ISSUE TITLE
Guarded recursive type does not pass positivity check

### ISSUE THREAD

**Issue by sgraf812** [agda/agda#6587]
I have the following type, following the [`LaterPrims`](https://github.com/agda/agda/blob/172366db528b28fb2eda03c5fc9804f2cdb1be18/test/Succeed/LaterPrims.agda) example without knowing exactly what I'm doing:

```agda
{-# OPTIONS --cubical --guarded #-}
module Test where

open import Agda.Primitive
open import Agda.Primitive.Cubical

module Prims where
  primitive
    primLockUniv : Set₁

open Prims renaming (primLockUniv to LockU) public

postulate
  Tick : LockU

▹_ : ∀ {l} → Set l → Set l
▹ A = (@tick x : Tick) -> A

data Domain : Set where
  d : ▹ (Domain -> Domain) -> Domain
```

This is rejected (tried 2.6.2.2 and master via the flake):

```
Checking Test (/home/sgraf/code/agda/test/Test.agda).
/home/sgraf/code/agda/test/Test.agda:18,1-19,37
Domain is not strictly positive, because it occurs
to the left of an arrow
in the type of the constructor d
in the definition of Domain.
```

Is this expected?

**Comment by andreasabel** [comment]
The positivity checker does not have knowledge of the _later_ modality, so it rejects this negative definition.
Is it safe to ignore negative occurrences under _later_?
Is there some literature for it?

**Comment by jespercockx** [comment]
The paragraph titled "Guarded recursive types with negative variance" on page 11 of https://arxiv.org/pdf/1611.09263.pdf seems to discuss this question explicitly:
> Guarded recursive types with negative variance. A key feature of guarded recursive types
are that they support negative occurrences of recursion variables. This is important for applications
to models of program logics [ 7].

**Comment by sgraf812** [comment]
Yes, I had hoped that this was one of the main points about *later* (although the main main point of course is a semantic instead of syntactic productivity check). 

Jesper gave an interesting reference that explicitly mentions negative occurrences that I hadn't found before.
I reflected a bit about it, re-read some papers and concluded that the positivity checker (well, some component of the type system) is probably correct in rejecting `Domain = |> (Domain -> Domain)`, because `|> Domain` does not appear verbatim on the "RHS". E.g., as a fixpoint: `Domain = μT.|>(T->T)`. Here, `λX.|>(X -> X)` does not have type `(|> A) -> A`. 
By contrast, `μT. (|> T -> |> T)` *should* be OK. (Note still that there is an inherent and annoying confusion about "type level" `|> T` and "kind level" `|> A` here. The papers I just skimmed used `λ(α:κ).T` as the clock-irrelevant term former for `|> A` instead of `|> T`, probably for that reason.)

Still,

```agda
data Domain2 : Set where
  d : (▹ Domain2 -> ▹ Domain2) -> Domain2

data Domain3 : Set where
  d : (▹ Domain3 -> Domain3) -> Domain3

data Domain4 : Set where
  d : ▹ (▹ Domain4 -> Domain4) -> Domain4
```

is rejected, too (on 2.6.2.2).

Edit: Actually, `Domain` might be OK, too, because of the isomorphism `λX.|>(X -> X) = λX.λ(α:κ).(X -> X) ~= λX.((λ(α:κ).X) -> (λ(α:κ).X)) = λX. |>X -> |>X` that I think holds. But honestly I can't claim that I completely understand the type systems involved.

**Comment by sgraf812** [comment]
Just to add to my wishlist in this issue, here's another example of a data type factoring that I think should be accepted by the positivity checker:

```agda
data Val : Set
data Domain5 : Set where
  d :  ▹ Val -> Domain5
data Val where
  fun : (Domain5 -> Domain5) -> Val
```

~~This kind of loop occurs in a more complex example of mine (my first Agda "program", apologies for the mess that I caused in anger) : 
https://gist.github.com/sgraf812/f00af45b1466d91e986cfbc631f66879#file-semantics-agda-L65~~

Edit: Well, ironically there actually *is* an unsound negative occurrence through `Dom -> T\infty -> Val -> Dom` that was not detected due to the deactivated positivity check :sweat_smile: 

... and this typechecks :) (yay), but I currently need to "break" the loop manually by deactivating the positivity check in one key place (the equivalent of `Domain5`) which I think is sound.

**Comment by sgraf812** [comment]
Just so that I don't lose track of where I found precedent for what is and what isn't allowed as a (guarded) recursive type, in Section 5.2 of https://dl.acm.org/doi/10.1145/3290317, they discuss the following type to model Milner's CCS, where any recursive occurrence of the type variable `X` under `μX.e` term former must be guarded by an *action* `a`, so `μX. a.X` is allowed, but not `μX. P ∥ X`. 

They use guarded recursive types to encode the guardedness condition, but use plain inductive recursive definitions anywhere else, e.g., within `P || P` term formers. So they have a "domain equation" like `CCS(n) = ... |>CCS(n) ... CCS(n) ...` which they solve with `CCS = fix X. μY. ... X ... Y ...`, where `fix` is the guarded fixed-point and `μ` is the least/inductive fixed-point operator.

Applied to `Domain3` above: `Domain3 = fix X. μ Y. (X -> Y)`. (Swapping args to `->` would of course *not* be OK.)

**Comment by sgraf812** [comment]
I recently learned that I can use guarded `fix` to define the corresponding data type from a signature functor. That is, the following module typechecks:

```
{-# OPTIONS --cubical --guarded #-}
module Test where

open import Cubical.Core.Everything
open import Cubical.Foundations.Everything
open import Agda.Primitive.Cubical

module Prims where
  primitive
    primLockUniv : Set₁

open Prims renaming (primLockUniv to LockU) public

postulate
  Tick : LockU

▹_ : ∀ {l} → Set l → Set l
▹ A = (@tick x : Tick) -> A

▸_ : ∀ {l} → ▹ Set l → Set l
▸ A = (@tick x : Tick) → A x

next : ∀ {l} {A : Set l} → A → ▹ A
next x _ = x

postulate
  dfix : ∀ {l} {A : Set l} → (▹ A → A) → ▹ A
  pfix : ∀ {l} {A : Set l} (f : ▹ A → A) → dfix f ≡ next (f (dfix f))

fix : ∀ {l} {A : Set l} → (▹ A → A) → A
fix f = f (dfix f)

data DomainF (d⁻ : ▹ Set) : Set where
  fun : (▸ d⁻ -> DomainF d⁻) -> DomainF d⁻

Domain : Set
Domain = fix DomainF

roll : Domain ≡ DomainF (next Domain)
roll i = DomainF (pfix DomainF i)
```

where `roll` can be used to unfold one layer of the signature functor. (Note the use of the black triangle ▸ in the signature functor, a "dependent tick". All postulates are from the "Guarded Cubical Agda Prelude" in [LaterPrims](https://github.com/agda/agda/blob/172366db528b28fb2eda03c5fc9804f2cdb1be18/test/Succeed/LaterPrims.agda).) 

I guess the OP then says: "It would be great if `roll` where somehow built-in the same way as it is for regular data types with positive recursion."
